runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperator.java

```diff
@@ -0,0 +1,233 @@
+              new SdkHarnessClient.RemoteOutputReceiver<WindowedValue<?>>() {
+                @Override
+                public Coder<WindowedValue<?>> getCoder() {
+                  return coder;
+                }
+
+                @Override
+                public FnDataReceiver<WindowedValue<?>> getReceiver() {
+                  return new FnDataReceiver<WindowedValue<?>>() {
+                    @Override
+                    public void accept(WindowedValue<?> input) throws Exception {
+                      logger.finer(String.format("Receiving value: %s", input));
+                      // TODO: Can this be called by multiple threads? Are calls guaranteed to
+                      // at least
+                      // be serial? If not, these calls may need to be synchronized.
+                      // TODO: If this needs to be synchronized, consider requiring immutable maps.
+                      synchronized (collectorLock) {
+                        // TODO: Get the correct output union tag from the corresponding output tag.
+                        // Plumb through output tags from process bundle descriptor.
+                        // NOTE: The output map is guaranteed to be non-empty at this point,
+                        // so we can
+                        // always grab index 0.
+                        // TODO: Plumb through TupleTag <-> Target mappings to get correct union tag
+                        outputManager.output(null, input);
+                      }
+                    }
+                  };
+                }
+              };
+      receiverBuilder.put(target, receiver);
+    }
+    Map<BeamFnApi.Target,
+            SdkHarnessClient.RemoteOutputReceiver<?>> receiverMap =
+            receiverBuilder.build();
+
+    SdkHarnessClient.ActiveBundle<InputT> bundle = processor.newBundle(receiverMap);
+    try (CloseableFnDataReceiver<WindowedValue<InputT>> inputReceiver = bundle.getInputReceiver()) {
+      logger.finer(String.format("Sending value: %s", element));
+      inputReceiver.accept(element);
+    }
+
+    // Await all outputs and active bundle completion. This is necessary because the Flink collector
+    // must not be accessed outside of mapPartition.
+    bundle.getOutputClients().values().forEach(client -> {
+      try {
+        client.awaitCompletion();
+      } catch (Exception e) {
+        throw new RuntimeException(e);
+      }
+    });
+    MoreFutures.get(bundle.getBundleResponse());
+  }
+
+  @Override
+  public void close() throws Exception {
+    // TODO: In what order should the client and session be closed? Should the session eventually
+    // own the client itself?
+    client.close();
+    client = null;
+    session.close();
+
+    super.close();
+  }
+
+  // TODO: remove single element bundle assumption
+  @Override
+  protected DoFnRunner<InputT, OutputT> createWrappingDoFnRunner(DoFnRunner<InputT, OutputT> wrappedRunner) {
```

Comment:
@aljoscha @bsidhom it looks as if the DoFnRunner override can be used for the SDK harness interaction w/o rewriting the full operator.